test/common/upstream/health_checker_impl_test.cc

```diff
@@ -2598,6 +2518,19 @@ class TestProdHttpHealthChecker : public ProdHttpHealthCheckerImpl {
 
 class ProdHttpHealthCheckerTest : public HttpHealthCheckerImplTest {
 public:
+  void allocHealthChecker(const std::string& yaml) {
```

Comment:
I will do so but hierarchy will change like below
- HealthCheckerTestBase
   - HttpHealthCheckerImplTest
      - ProdHttpHealthCheckerTest
   - TcpHealthCheckerImplTest
   - GrpcHealthCheckerImplTestBase
     - GrpcHealthCheckerImplTest
     - BadResponseGrpcHealthCheckerImplTest

Is it okay?

Reply:
Per the above comment, I do not think it makes sense to have ProdHttpHealthCheckerTest inherit from HttpHealthCheckerImplTest. This becomes clearer once you move the previously-mentioned member variables out of HttpHealthCheckerImplTest and into HealthCheckerTestBase.

Once you do that move, which will simplify all the test classes, there will be nothing in common between ProdHttpHealthCheckerTest and HttpHealthCheckerImplTest and it will be clearer they should be siblings and not base/derived. I could be wrong about this but this is what I got out of reading their member-variable lists.

I think the main issue with the BadResponse* tests is that they are parameterized and the GrpcHealthCheckerImplTest is not, but if you can make that work then I'm good with that structure. The main litmus test for me is whether methods in the derived class uses anything declared in the parent class.